src/nec-usbatermfw.c

```diff
@@ -0,0 +1,305 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (C) 2024 INAGAKI Hiroshi <musashino.open@gmail.com>
+ */
+
+#include <byteswap.h>
+#include <endian.h>
+#include <errno.h>
+#include <libgen.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#define FWHDR_LEN		0x20
+#define BLKHDR_LEN		0x18
+
+#define BLKHDR_FLAGS_GZIP	0x80000000
+#define BLKHDR_FLAGS_EXEC	0x00000002
+#define BLKHDR_FLAGS_SYS	0x00000001
+#define BLKHDR_DEF_LOADADDR	0x80040000
+#define BLKHDR_DEF_ENTRYP	BLKHDR_DEF_LOADADDR
+
+#define FILETYPE_LEN		0x20
+
+#define PADBLK_LEN		0x4
+#define DATABLK_CNT_MAX		2
+
+struct header
+{
+	uint32_t flags;
+	uint32_t totlen;
+	uint32_t hdrlen;
+	uint32_t cksum;
+	uint32_t loadaddr;
+	uint32_t entryp;
+} __attribute__ ((packed));
+
+static char *progname;
+FILE *outbin;
+int nblk = 0;
+
+static void usage(void)
+{
+	printf("Usage: %s <output> [OPTIONS...]\n", progname);
+	printf("\n"
+	       "Options:\n"
+	       "  -t <type>       set firmware type to <type>\n"
+	       "  -f <flags>      set data flags to <flags>\n"
+	       "  -a <loadaddr>   set data load address to <loadaddr>\n"
+	       "  -e <entry>      set data entry point to <entry>\n"
+	       "  -d <file>       read input data from file <file> (max. %dx)\n",
+	       DATABLK_CNT_MAX);
+}
+
+static int strtou32(char *arg, uint32_t *val)
+{
+	char *endptr = NULL;
+
+	errno = 0;
+	*val = strtoul(arg, &endptr, 0);
+	return (errno || (endptr && *endptr)) ? -EINVAL : 0;
+}
+
+static int write_data(const char *buf, size_t length, uint32_t *cksum)
+{
+	size_t padlen = 0;
+	size_t pos;
+
+	if (length % PADBLK_LEN)
+		padlen = PADBLK_LEN - length % PADBLK_LEN;
+
+	if (fwrite(buf, 1, length, outbin) != length) {
+		fprintf(stderr, "Couldn't write %zu bytes\n", length);
+		return -EACCES;
+	}
+
+	if (cksum)
+		for (pos = 0; pos < length; pos += 2) {
+			if (length - pos == 1)
+				*cksum += le16toh(buf[pos]);
+			else
+				*cksum += le16toh(*(uint16_t *)(buf + pos));
+		}
+
+	if (padlen) {
+		int i;
+		for (i = 0; i < padlen; i++)
+			fputc(0x0, outbin);
+	}
+
+	return padlen;
+}
+
+static int write_blockheader(uint32_t flags,
+			     uint32_t loadaddr, uint32_t entryp,
+			     size_t datalen, int padlen, uint32_t cksum)
+{
+	struct header hdr;
+	uint16_t *cur = (uint16_t *)&hdr;
+	char buf[0x20];
+
+	flags <<= 16;
+	flags |= (~flags & 0xffff0000) >> 16;
+
+	hdr.flags = htobe32(flags);
+	hdr.totlen = htobe32(BLKHDR_LEN + datalen);
+	hdr.hdrlen = htobe32(BLKHDR_LEN);
+	hdr.cksum = 0;
+	hdr.loadaddr = htobe32(loadaddr);
+	hdr.entryp = htobe32(entryp);
+
+	for (cur += 2; cur - (uint16_t *)&hdr < BLKHDR_LEN / 2; cur++) {
+		cksum += htole16(*cur);
```

Comment:
Isn't this:
```suggestion
		cksum += be16toh(*cur);
```

Reply:
No, it should be `htole16` (or `le16toh`).

```
IPL:HARD-RESET
memory test ... ok
flinstall OK

boot version: 1.0.0

PKG "Orthos-HP"
CPU "Atheros/AR9344"
Revision 212.1 CLOCK 560.0 MHz
HEAP MEMORY 262144 bytes
Boot:InitFilesystem
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
post start
DRAM Test ..... PASS
Timer test .... PASS
FLASH CHKSUM .. PASS
LAN Test ..... 
LoopBack count 0 .. 
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
active ch:1
phy0.1: "Atheros/AR8327(SW)"
phy0.0: "Atheros/AR8327(phy)"
active ch:0
PASS
Package TYPE = 0x01
post success

welcome to boot console
Password:
boot> tftpd
tftpd start 192.168.0.1
boot> start tftp load openwrt-ath79-tiny-nec_wg600hp-i         <-- be16toh()
end tftp load length = 5578496
start memory load ... 
load error [5]

boot> start tftp load openwrt-ath79-tiny-nec_wg600hp-i         <-- htole16()
end tftp load length = 5578496
start memory load ... 
memory load complete
  begin  : 0x80040000
  length : 5575740
  startup: 0x80040000

```